File: csharp/lib/Property/WhiteBalanceProperty.cs
/*
 * SPDX-FileCopyrightText: 2017-2022 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

using MessagePack;

namespace SensCord
{
    /// <summary>
    /// Structure for the image of the camera exposure.
    /// </summary>
    [MessagePackObject]
    public class WhiteBalanceProperty : IBaseProperty
    {
        /// <summary>
        /// Property key.
        /// </summary>
        public static readonly string ConstKey = "white_balance_property";

        /// <summary>
        /// Property key.
        /// </summary>
        [IgnoreMember]
        public string Key { get; } = ConstKey;

        /// <summary>
        /// Mode of white balance.
        /// </summary>
        [Key("mode")]
        public string Mode { get; set; } = string.Empty;

        /// <summary>
        /// To string.
        /// </summary>
        /// <returns>A string that represents the current object.</returns>
        public override string ToString()
        {
            return string.Format($"{0}({1})", this.Key, this.Mode);
        }
    }

    /// <summary>
    /// White balance modes.
    /// </summary>
    public static partial class WhiteBalanceMode
    {
        /// <summary>
        /// Auto.
        /// </summary>
        public static readonly string Auto = "auto";

        /// <summary>
        /// Manual.
        /// </summary>
        public static readonly string Manual = "manual";
    }
}

File: java/senscord/src/senscord_properties/WhiteBalanceProperty.java
/*
 * SPDX-FileCopyrightText: 2019 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package senscord_properties;

import com.sun.jna.Native;
import com.sun.jna.Structure;
import java.util.Arrays;
import java.util.List;

/**
 * WhiteBalanceProperty.
 * 
 * @see "senscord::WhiteBalanceProperty"
 */
public class WhiteBalanceProperty extends Structure {
    public byte[] mode;

    public static class ByReference extends WhiteBalanceProperty implements Structure.ByReference {
    }

    @Override
    protected List<String> getFieldOrder() {
        return Arrays.asList("mode");
    }

    public WhiteBalanceProperty() {
        super();
        mode = new byte[64];
    }

    /**
     * Exposure metering modes.
     */
    /* Metering mode : None (Disable) */
    public static interface Mode {
        /**
         * Auto.
         */
        public static final String SENSCORD_WHITE_BALANCE_AUTO = "auto";

        /**
         * Manual.
         */
        public static final String SENSCORD_WHITE_BALANCE_MANUAL = "manual";

    }

    /**
     * Get white balance mode.
     * 
     * @return white balance mode from {@link Mode}
     */
    public String GetMode() {
        return Native.toString(mode);
    }

    /**
     * Set white balance mode.
     * 
     * @param mode white balance mode form {@link Mode}
     */
    public void SetMode(String mode) {
        Arrays.fill(this.mode, (byte) 0);
        System.arraycopy(mode.getBytes(), 0, this.mode, 0, mode.length());
    }

}
